File: examples/java/projector-log-transaction/src/main/java/dev/angzarr/examples/projector/TransactionLogProjectorService.java
package dev.angzarr.examples.projector;

import com.google.protobuf.Empty;
import io.grpc.stub.StreamObserver;
import dev.angzarr.EventBook;
import dev.angzarr.Projection;
import dev.angzarr.ProjectorCoordinatorGrpc;

/**
 * gRPC service for transaction log projector.
 */
public class TransactionLogProjectorService extends ProjectorCoordinatorGrpc.ProjectorCoordinatorImplBase {

    private final TransactionLogProjector projector;

    public TransactionLogProjectorService(TransactionLogProjector projector) {
        this.projector = projector;
    }

    @Override
    public void handle(EventBook request, StreamObserver<Empty> responseObserver) {
        projector.logEvents(request);
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void handleSync(EventBook request, StreamObserver<Projection> responseObserver) {
        projector.logEvents(request);
        responseObserver.onNext(Projection.getDefaultInstance());
        responseObserver.onCompleted();
    }
}

File: examples/csharp/ProjectorLogTransaction/src/TransactionLogProjectorService.cs
using Angzarr;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;

namespace Angzarr.Examples.Projector;

public class TransactionLogProjectorService : ProjectorCoordinator.ProjectorCoordinatorBase
{
    private readonly ITransactionLogProjector _projector;

    public TransactionLogProjectorService(ITransactionLogProjector projector)
    {
        _projector = projector;
    }

    public override Task<Empty> Handle(EventBook request, ServerCallContext context)
    {
        _projector.LogEvents(request);
        return Task.FromResult(new Empty());
    }

    public override Task<Projection> HandleSync(EventBook request, ServerCallContext context)
    {
        _projector.LogEvents(request);
        // Log projector doesn't produce a projection
        return Task.FromResult<Projection>(null!);
    }
}
